src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,608 @@
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis.ConvertLinq;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Simplification;
+
+namespace Microsoft.CodeAnalysis.CSharp.ConvertLinq
+{
+    internal sealed class CSharpConvertLinqQueryToLinqMethodProvider : AbstractConvertLinqProvider
+    {
+        private static readonly TypeSyntax VarNameIdentifier = SyntaxFactory.IdentifierName("var");
+
+        protected override IAnalyzer CreateAnalyzer(SemanticModel semanticModel, CancellationToken cancellationToken)
+            => new CSharpAnalyzer(semanticModel, cancellationToken);
+
+        private sealed class CSharpAnalyzer : AnalyzerBase<QueryExpressionSyntax, SyntaxNode, SyntaxNode>
+        {
+            public CSharpAnalyzer(SemanticModel semanticModel, CancellationToken cancellationToken)
+                : base(semanticModel, cancellationToken)
+            {
+            }
+
+            protected override string Title => CSharpFeaturesResources.Convert_LINQ_query_to_foreach;
+
+            protected override bool TryConvert(QueryExpressionSyntax source, out DocumentUpdate documentUpdate)
+            {
+                if (_semanticModel.GetDiagnostics(source.Span, _cancellationToken).Any() ||
+                    source.DescendantTrivia().Any(trivia => trivia.MatchesKind(
+                        SyntaxKind.SingleLineCommentTrivia,
+                        SyntaxKind.MultiLineCommentTrivia,
+                        SyntaxKind.MultiLineDocumentationCommentTrivia) ||
+                        source.ContainsDirectives))
+                {
+                    documentUpdate = default;
+                    return false;
+                }
+
+                return TryConvertInternal(source, source.Parent, out documentUpdate);
+            }
+
+            private bool TryConvertInternal(QueryExpressionSyntax source, SyntaxNode parent, out DocumentUpdate documentUpdate)
+            {
+                switch (parent.Kind())
+                {
+                    case SyntaxKind.ReturnStatement:
+                        return TryConvertIfInReturnStatement(source, (ReturnStatementSyntax)parent, out documentUpdate);
+                    case SyntaxKind.ForEachStatement:
+                        return TryConvertIfInForEach(source, (ForEachStatementSyntax)parent, out documentUpdate);
+                    case SyntaxKind.EqualsValueClause:
+                    case SyntaxKind.SimpleAssignmentExpression:
+                        return TryReplaceWithLocalFunction(source, parent, out documentUpdate);
+                    case SyntaxKind.ParenthesizedExpression:
+                        return TryConvertInternal(source, parent.Parent, out documentUpdate);
```

Comment:
interesting. this exists to just recursive walk upwards, unwinding parens.  We have the WalkUpParentheses helper for this purpose.  I think it would be clearer to just use that and not be recursive.

Reply:
Thank you! I thought about some other options when we need to skip some node. If there are no such cases, let us use WalkUpParentheses and remove the parent as a parameter.

---
In reply to: [176297119](https://github.com/dotnet/roslyn/pull/25362#discussion_r176297119) [](ancestors = 176297119)

Reply:
Agreed! :)